garden_ai/containers.py

```diff
@@ -40,15 +41,19 @@ class DockerBuildFailure(docker.errors.BuildError):
 def handle_docker_errors(func):
     """
     This decorator catches common classes of Docker errors and sends recommendations for how to fix them up the callstack.
-    Right now it just catches exceptions that are thrown when the user can't run Docker at all.
-    If it's in this class of error, it raises a DockerStartFailure.
+    If the user can't run Docker at all, it raises a DockerStartFailure.
+    If the user runs into a BuildError, it tries to remove any dangling intermediate images before re-raising the exception.
     Otherwise the normal Docker exception is still raised.
     """
 
     @functools.wraps(func)
     def wrapper(*args, **kwargs):
         try:
             return func(*args, **kwargs)
+        except docker.errors.BuildError:
+            client = get_docker_client()
+            cleanup_dangling_images(client)
```

Comment:
I'm skeptical of how this PR removes _all_ dangling images from the user's machine here and at the start of the prune command. If we knew we were the only application using Docker on their machine I could see it. But I think it's unexpected/surprising behavior that could interfere with other applications.

At most I think we should include a flag in the prune command that enables this. I think we could even get away with not including it at all and pointing users to the existing `docker system prune` command

Reply:
Yeah glad you said something, I was on the fence / don't love the risk of overstepping by cleaning too much. The situation I'm worried about is one where a failed build leaves an untagged image behind which references one of the large base images, then `garden-ai docker prune` cleans out every base image except that one for no obvious reason. 

I could probably filter this exception handling down to "dangling images created since the start of the build process" and skip it in the actual command if that seems sane, or could try and limit it to images that share layers with known garden images